pkg/advisory/alias_finder.go

```diff
@@ -0,0 +1,176 @@
+package advisory
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"net/url"
+	"os"
+	"time"
+
+	"github.com/samber/lo"
+)
+
+type AliasFinder interface {
+	CVEForGHSA(ctx context.Context, ghsaID string) (string, error)
+	GHSAsForCVE(ctx context.Context, cveID string) ([]string, error)
+}
+
+type HTTPAliasFinder struct {
+	client *http.Client
+}
+
+func NewHTTPAliasFinder(client *http.Client) *HTTPAliasFinder {
+	return &HTTPAliasFinder{
+		client: client,
+	}
+}
+
+func (f *HTTPAliasFinder) CVEForGHSA(ctx context.Context, ghsaID string) (string, error) {
+	requestPath := fmt.Sprintf("/advisories/%s", url.PathEscape(ghsaID))
+	respBody, err := f.gitHubAPIGet(ctx, requestPath, nil)
+	if err != nil {
+		return "", err
+	}
+	defer respBody.Close()
+
+	var ghsa githubSecurityAdvisoryResponse
+	if err := json.NewDecoder(respBody).Decode(&ghsa); err != nil {
+		return "", err
+	}
+
+	return ghsa.CVEID, nil
+}
+
+func (f *HTTPAliasFinder) GHSAsForCVE(ctx context.Context, cveID string) ([]string, error) {
+	respBody, err := f.gitHubAPIGet(
+		ctx,
+		"/advisories",
+		map[string]string{"cve_id": cveID},
+	)
+	if err != nil {
+		return nil, err
+	}
+	defer respBody.Close()
+
+	var ghsas []githubSecurityAdvisoryResponse
+	if err := json.NewDecoder(respBody).Decode(&ghsas); err != nil {
+		return nil, err
+	}
+
+	cveIDs := lo.Map(ghsas, func(ghsa githubSecurityAdvisoryResponse, _ int) string {
+		return ghsa.GHSAID
+	})
+
+	return cveIDs, nil
+}
+
+func (f *HTTPAliasFinder) gitHubAPIGet(ctx context.Context, requestPath string, queryParameters map[string]string) (io.ReadCloser, error) {
+	u, err := url.Parse("https://api.github.com")
+	if err != nil {
+		return nil, err
+	}
+	u.Path = requestPath
+	if queryParameters != nil {
+		q := u.Query()
+		for k, v := range queryParameters {
+			q.Set(k, v)
+		}
+		u.RawQuery = q.Encode()
+	}
+
+	req, err := http.NewRequestWithContext(ctx, http.MethodGet, u.String(), nil)
+	if err != nil {
+		return nil, err
+	}
+
+	req.Header = http.Header{
+		"Accept":               []string{"application/vnd.github+json"},
+		"X-GitHub-Api-Version": []string{"2022-11-28"},
+	}
+
+	if token := os.Getenv("GITHUB_TOKEN"); token != "" {
+		req.Header.Set("Authorization", fmt.Sprintf("Bearer %s", token))
+	}
+
+	resp, err := f.client.Do(req)
+	if err != nil {
+		return nil, err
+	}
+	if resp.StatusCode != http.StatusOK {
+		var msg []byte
+		if resp.Body != nil {
+			//nolint:errcheck // This is a best-effort attempt to read the body.
+			msg, _ = io.ReadAll(resp.Body)
+		}
+		return nil, fmt.Errorf("unexpected status code %d, body: %s", resp.StatusCode, string(msg))
+	}
+
+	return resp.Body, nil
+}
+
+type githubSecurityAdvisoryResponse struct {
```

Comment:
Not really actionable for now, but it would be neat to see most of the logic talking to GitHub incorporated into https://github.com/google/go-github so we can reuse it later. We'll need more integration with GitHub and having a structured way of fetching interacting with GitHub would be helpful long term.

Looks like Billy contributed a bunch to go-github repo

Reply:
Good callout, I actually started with trying to use https://github.com/google/go-github, but then I saw they don't support the endpoint I need just yet. There's an open issue for it: https://github.com/google/go-github/issues/2851